src/algebra.jl

```diff
@@ -37,7 +37,7 @@ Base.:(+)(xs::AbstractBlock...) = Sum(xs...)
 Base.:(*)(xs::AbstractBlock...) = chain(Iterators.reverse(xs)...)
 Base.:(/)(A::AbstractBlock, x::Number) = (1/x)*A
 # reduce
-Base.sum(a::AbstractBlock{N}, blocks::AbstractBlock{N}...) where N = Sum(a, blocks...)
+Base.sum(blocks::AbstractArray{T}) where {N, T<:AbstractBlock{N}} = Sum(blocks...)
```

Comment:
I think we don't need `sum` since `+` is already defined? We could just the what's in Base, there is no  need to specialize this and `prod`, just delete them is fine?

Reply:
`sum` will  produce `((a+b)+c)+d` by default? Now we don't have automatic optimization, the nested `+` will be ugly.

Reply:
But I am ok to change if you think this is nessesary. This interface is same as `Sum`.

Reply:
So change the name to `AddBlock`?

Reply:
Yeah, kind of. But I don't like `XXBlock` convention now after we split things to packages since you have to type `YaoBlocks.XXBlocks`, there is a `Block` in the module name already which is pretty explicit already, so I would say just `Add` or `Plus` is fine. 

One reason of this convention change is we might need to remove some extra functions like `chain` and encourage people to use the constructor directly in the future to simplify things (after the auto currying feature of `_`), currying argument is the only reason we are still using functions. So the way of writing the same thing can be eliminated (you won't want to have many alternatives to implement the exact same thing)

Reply:
```julia
julia> Base.:+(x::AbstractBlock, args...) = Sum([x, args...])

julia> sum([X, Y, Z])
nqubits: 1
+
├─ +
│  ├─ X gate
│  └─ Y gate
└─ Z gate
```

Does not work? Do you mean defining `+(::AbstractBlock, ::Sum)`?

Reply:
umm, strange, what about just define `+` then? and leave `sum` as it is. We could have `eliminate nested` enabled in `Sum` constructor.

Reply:
I could do the elimination part in `Sum` constructor, you can just define the `+` and delete `sum` here, and I'll merge this PR.